src/output.c

```diff
@@ -354,6 +368,19 @@ new_output_notify(struct wl_listener *listener, void *data)
 	server->pending_output_layout_change--;
 	do_output_layout_change(server);
 	seat_output_layout_changed(&output->server->seat);
+
+	if (server->fallback_output && wl_list_length(&server->outputs) > 1) {
+		/* libwayland has a bug when a global is created and
+		 * immediately destroyed, as clients don't have enough time
+		 * to bind it, so the virtual output is destroyed on a timer.
+		 */
```

Comment:
This seems like a pretty hacky workaround and may or may not work depending on the clients responding within 1 sec..

Is there a issue for that at wlroots?

Reply:
I don't know - I am basing it on the code at https://github.com/WayfireWM/wayfire/blob/master/src/core/output-layout.cpp#L1520, and certainly destroying the virtual output immediately caused segfaults under certain circumstances; specifically when starting up with a display already connected.

Reply:
I played around a bit with an implementation without timers so we only store the `wlr_output` of the fallback output and decide based on that pointer. That mostly works but is a bit annoying as well because wlroots will emit the new output signal before returning from `wlr_headless_add_output()`.

However, I am not sure if adding and removing the fallback output is the best option here. Maybe we could instead always add the fallback output (before the headless backend has been started to not run into the issue described above) and then keep the fallback output around but disable as soon as some other output is detected + enable when the last output is gone.

Reply:
I did try disabling and re-enabling instead of destroying and creating, but it means you always see the headless output in tools like wlr-randr, which is potentially confusing to the user. I have modified arandr to provide a graphical front end to wlr-randr as well as xrandr, and just disabling the headless output would mean it would appear in arandr and the user could then enable and resize it when it wasn't actually valid.

Reply:
If you can, that would be great - I spent a couple of days trying everything I could think of to get rid of it, but I suspect you'll have more sensible ideas than I did!

Reply:
Feel free to give https://github.com/Consolatis/labwc/tree/feature/virtual_fallback_output a try.

I could only test it nested which sadly doesn't allow outputs to re-appear after being destroyed. The name of the fallback output is read from the environment, e.g. `LABWC_FALLBACK_OUTPUT=HEADLESS-fallback labwc`.

Reply:
Sorry, that doesn't start the desktop for me if I boot with an external display connected; it just falls back to the greeter every time. I can see the headless output if I boot with no external display connected, and if I connect an external display once booted to headless, that works, but initial boot with external display connected fails. 

BTW, that was exactly the problem I had before adding the timeout on the destroy - there seems to be some sort of startup race condition between the initial creation of the fallback output and its being destroyed when a real output is detected, and I suspect you are seeing the same problem. I was getting two separate destroy notifications, and the second one had bad data resulting in a segfault.

Also - it doesn't seem to make any difference whether I set the environment variable, or what I set it to - I always get a fallback output called "1" whether the variable is set or not.

Reply:
> Also - it doesn't seem to make any difference whether I set the environment variable, or what I set it to - I always get a fallback output called "1" whether the variable is set or not.

As the name comes from the `LABWC_FALLBACK_OUTPUT` env var it seems be set somewhere (maybe in `~/.config/labwc/environment`?). You could try to `env | grep -i labwc` to verify if its set or not.

> Sorry, that doesn't start the desktop for me if I boot with an external display connected; it just falls back to the greeter every time. I can see the headless output if I boot with no external display connected, and if I connect an external display once booted to headless, that works, but initial boot with external display connected fails.

I'll try to test on another system with DRM output later.

> BTW, that was exactly the problem I had before adding the timeout on the destroy - there seems to be some sort of startup race condition between the initial creation of the fallback output and its being destroyed when a real output is detected, and I suspect you are seeing the same problem. I was getting two separate destroy notifications, and the second one had bad data resulting in a segfault.

At least when running nested this seems to work fine for me, however it might be that the wayland backend is initialized first and then the headless backend while what you are experiencing is headless being initialized first and then the DRM one.

I decoupled the creation / destruction of the fallback output from the `new_output_notify/destroy` handler and moved it into the output layout change event, that way we should not execute anything twice (and also catch the case with disabled or DRM leased outputs).

Reply:
Had a chance to test on a DRM setup and could a) reproduce the crash and b) hopefully fix it.

The reason for the crash was that the output was destroyed while in the new_output handler of itself, e.g. the original issue.

That in turn happened because we unconditionally added the fallback output when starting up, in the DRM case this means that the actual DRM outputs may be discovered earlier than the fallback output when starting up the backends. Thus we ended up in a situation where we already added the fallback output but get the notification only when the backends start up at which point there are already the DRM outputs present. Which in turn means that we try to destroy it while its being added and wlroots really doesn't like that.

The way I tackled it is that we now a) move the creation of the fallback output \*after\* the backends have already been started and b) guard it with a check that there are indeed no other outputs available at that point.

The branch test branch is updated and could need some more testing.

Reply:
Yes, that was what I saw - the problem arose when I created the headless output and then it was immediately destroyed when the real output was detected; I couldn't work out how to conditionally create it only after waiting to see if a real output appeared. So I think that will indeed fix the crash - I'll test some more in the morning.

Reply:
OK, just tried this again and I can't repeat that problem - even without LABWC_FALLBACK_OUTPUT being set, I am still able to connect to the physical display. Not sure what was going wrong with this last week, but it seems to be ok now - must have been finger trouble my end. Apologies for the spurious problem report!

Reply:
Is it possible to get your branch accepted as a PR into labwc? It seems to work pretty well to me.

Reply:
Sure, its too risky to merge before 0.7.1 though, once the release is done I'll open a PR.

Reply:
Created the PR:

- #1586

Refactored it slightly, testing would be appreciated.